src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts

```diff
@@ -0,0 +1,747 @@
+              const n = Math.floor(v / textureSize[i]);
+              v = v - n * textureSize[i];
+              return (n & 1) !== 0 ? textureSize[i] - v - 1 : v;
+            }
+            case 'repeat':
+              return v - Math.floor(v / textureSize[i]) * textureSize[i];
+            default:
+              unreachable();
+          }
+        });
+        const v = load(c);
+        ss.push(v);
+        for (const component of rep.componentOrder) {
+          out[component] = (out[component] ?? 0) + v[component]! * sample.weight;
+        }
+      }
+
+      return out;
+    }
+    case 'textureLoad': {
+      return load(toArray(call.coords!));
+    }
+  }
+}
+
+export async function putDataInTextureThenDrawAndCheckResults<T extends Dimensionality>(
+  device: GPUDevice,
+  texture: Texture,
+  sampler: GPUSamplerDescriptor,
+  calls: TextureCall<T>[]
+) {
+  const results = await doTextureCalls(device, texture, sampler, calls);
+  const errs: string[] = [];
+  const rep = kTexelRepresentationInfo[texture.texels.format];
+  for (let callIdx = 0; callIdx < calls.length; callIdx++) {
+    const call = calls[callIdx];
+    const got = results[callIdx];
+    const expect = expected(call, texture, sampler);
+
+    const gULP = rep.bitsToULPFromZero(rep.numberToBits(got));
+    const eULP = rep.bitsToULPFromZero(rep.numberToBits(expect));
+    for (const component of rep.componentOrder) {
+      const g = got[component]!;
+      const e = expect[component]!;
+      const absDiff = Math.abs(g - e);
+      const ulpDiff = Math.abs(gULP[component]! - eULP[component]!);
+      const relDiff = absDiff / Math.max(Math.abs(g), Math.abs(e));
+      if (ulpDiff > 3 && relDiff > 0.03) {
+        const desc = describeTextureCall(call);
+        errs.push(`component was not as expected:
+      call: ${desc}
+ component: ${component}
+       got: ${g}
+  expected: ${e}
+  abs diff: ${absDiff.toFixed(4)}
+  rel diff: ${(relDiff * 100).toFixed(2)}%
+  ulp diff: ${ulpDiff}
+  sample points:
+`);
+        const expectedSamplePoints = [
+          'expected:',
+          ...(await identifySamplePoints(texture.descriptor, (texels: TexelView) => {
+            return Promise.resolve(
+              expected(call, { texels, descriptor: texture.descriptor }, sampler)
+            );
+          })),
+        ];
+        const gotSamplePoints = [
+          'got:',
+          ...(await identifySamplePoints(
+            texture.descriptor,
+            async (texels: TexelView) =>
+              (
+                await doTextureCalls(device, { texels, descriptor: texture.descriptor }, sampler, [
+                  call,
+                ])
+              )[0]
+          )),
+        ];
+        errs.push(layoutTwoColumns(expectedSamplePoints, gotSamplePoints).join('\n'));
+        errs.push('', '');
+      }
+    }
+  }
+
+  return errs.length > 0 ? new Error(errs.join('\n')) : undefined;
+}
+
+export async function identifySamplePoints(
```

Comment:
For this function in particular, also please add a comment somewhere in the middle so it's easier to tell where the "interesting" part finishes and the string construction starts. Or consider splitting the function in two.

Also maybe put sample output in a comment so the reader has a picture of what the code is attempting to do.

```ts
// component was not as expected:
//       call: textureSample(texture: T, sampler: S, coords: vec2(1.125, 0.21875))
//  component: R
//        got: 0.4960784316062927
//   expected: 0.496078431372549
//   abs diff: 0.0000
//   rel diff: 0.00%
//   ulp diff: 1
//   sample points:
// 
// expected:                            | got:
//     0   1   2   3   4   5   6   7    |     0   1   2   3   4   5   6   7 
//   ┌───┬───┬───┬───┬───┬───┬───┬───┐  |   ┌───┬───┬───┬───┬───┬───┬───┬───┐
// 0 │   │   │   │   │   │   │   │   │  | 0 │   │   │   │   │   │   │   │   │
//   ├───┼───┼───┼───┼───┼───┼───┼───┤  |   ├───┼───┼───┼───┼───┼───┼───┼───┤
// 1 │   │   │   │   │   │   │   │ a │  | 1 │   │   │   │   │   │   │   │ a │
//   ├───┼───┼───┼───┼───┼───┼───┼───┤  |   ├───┼───┼───┼───┼───┼───┼───┼───┤
// 2 │   │   │   │   │   │   │   │ b │  | 2 │   │   │   │   │   │   │   │ b │
//   ├───┼───┼───┼───┼───┼───┼───┼───┤  |   ├───┼───┼───┼───┼───┼───┼───┼───┤
// 3 │   │   │   │   │   │   │   │   │  | 3 │   │   │   │   │   │   │   │   │
//   ├───┼───┼───┼───┼───┼───┼───┼───┤  |   ├───┼───┼───┼───┼───┼───┼───┼───┤
// 4 │   │   │   │   │   │   │   │   │  | 4 │   │   │   │   │   │   │   │   │
//   ├───┼───┼───┼───┼───┼───┼───┼───┤  |   ├───┼───┼───┼───┼───┼───┼───┼───┤
// 5 │   │   │   │   │   │   │   │   │  | 5 │   │   │   │   │   │   │   │   │
//   ├───┼───┼───┼───┼───┼───┼───┼───┤  |   ├───┼───┼───┼───┼───┼───┼───┼───┤
// 6 │   │   │   │   │   │   │   │   │  | 6 │   │   │   │   │   │   │   │   │
//   ├───┼───┼───┼───┼───┼───┼───┼───┤  |   ├───┼───┼───┼───┼───┼───┼───┼───┤
// 7 │   │   │   │   │   │   │   │   │  | 7 │   │   │   │   │   │   │   │   │
//   └───┴───┴───┴───┴───┴───┴───┴───┘  |   └───┴───┴───┴───┴───┴───┴───┴───┘
// a: at: [7, 1], weights: [R: 0.75000] | a: at: [7, 1], weights: [R: 0.75000]
// b: at: [7, 2], weights: [R: 0.25000] | b: at: [7, 2], weights: [R: 0.25000]
```

Reply:
done